File: plugins/presentation-skill/skills/presentation-skill/scripts/pulse_theme.py
#!/usr/bin/env python3
"""Reads pulse/theme.json for the Python half of the skill.

The JS renderer has the same loader in templates/pptxgenjs/pulse_theme.js.
Both read the one file a designer edits, so neither side carries a colour,
font, size or margin of its own.

Run as a script to validate the file:

    python scripts/pulse_theme.py --check
"""

from __future__ import annotations

import argparse
import json
import re
import sys
from pathlib import Path
from typing import Any

ROOT = Path(__file__).resolve().parent.parent
THEME_PATH = ROOT / "pulse" / "theme.json"

# theme.json inches -> renderer inches. The theme is written in the units a
# designer reads off PowerPoint (13.333in slide); the renderer authors at 10in.
SCALE = 4 / 3

HEX = re.compile(r"^[0-9a-fA-F]{6}$")

REQUIRED_COLORS = ("ink", "ink_muted", "bg", "surface", "accent", "line")
TYPE_ORDER = ("micro", "caption", "secondary", "body", "subtitle", "title", "hero")
REQUIRED_GRID = (
    "margin", "title_y", "title_h", "subtitle_y", "subtitle_h",
    "content_y", "content_bottom", "gutter",
)


class ThemeError(Exception):
    pass


def read_theme(path: Path | None = None) -> dict[str, Any]:
    file = Path(path) if path else THEME_PATH
    try:
        raw = file.read_text(encoding="utf-8")
    except OSError as exc:
        raise ThemeError(f"Не удалось прочитать {file}: {exc}") from exc
    try:
        return json.loads(raw)
    except json.JSONDecodeError as exc:
        raise ThemeError(
            f"{file} — некорректный JSON: {exc}\n"
            "Частая причина: лишняя запятая перед закрывающей скобкой."
        ) from exc


def validate_theme(theme: dict[str, Any]) -> list[str]:
    """Return human-readable problems; empty list means the file is good."""
    problems: list[str] = []

    colors = theme.get("colors")
    if not isinstance(colors, dict):
        problems.append('корень: нет обязательного ключа "colors"')
    else:
        for key in REQUIRED_COLORS:
            if key not in colors:
                problems.append(f'colors: нет обязательного ключа "{key}"')
            elif not HEX.match(str(colors[key])):
                problems.append(
                    f'colors.{key} = "{colors[key]}" — нужен HEX из шести '
                    "символов без решётки, например 005AFE"
                )

    fonts = theme.get("fonts")
    if not isinstance(fonts, dict):
        problems.append('корень: нет обязательного ключа "fonts"')
    else:
        for key in ("heading", "body"):
            if not str(fonts.get(key, "")).strip():
                problems.append(f"fonts.{key} пустой или отсутствует")

    scale = theme.get("type_scale_pt")
    if not isinstance(scale, dict):
        problems.append('корень: нет обязательного ключа "type_scale_pt"')
    else:
        for key in TYPE_ORDER:
            if key not in scale:
                problems.append(f'type_scale_pt: нет обязательного ключа "{key}"')
                continue
            try:
                if float(scale[key]) <= 0:
                    raise ValueError
            except (TypeError, ValueError):
                problems.append(
                    f'type_scale_pt.{key} = "{scale[key]}" — нужно число больше нуля'
                )
        for lo_key, hi_key in zip(TYPE_ORDER, TYPE_ORDER[1:]):
            try:
                lo, hi = float(scale[lo_key]), float(scale[hi_key])
            except (KeyError, TypeError, ValueError):
                continue
            if hi < lo:
                problems.append(
                    f'type_scale_pt: "{hi_key}" ({hi}) меньше "{lo_key}" ({lo}) — '
                    "шкала должна расти от micro к hero"
                )

    grid = theme.get("grid_in")
    if not isinstance(grid, dict):
        problems.append('корень: нет обязательного ключа "grid_in"')
    else:
        for key in REQUIRED_GRID:
            if key not in grid:
                problems.append(f'grid_in: нет обязательного ключа "{key}"')
                continue
            try:
                if float(grid[key]) < 0:
                    raise ValueError
            except (TypeError, ValueError):
                problems.append(
                    f'grid_in.{key} = "{grid[key]}" — нужно число не меньше нуля (дюймы)'
                )
        try:
            width = float(theme.get("slide", {}).get("width_in", 13.333))
            if float(grid["margin"]) * 2 >= width:
                problems.append(
                    f'grid_in.margin = {grid["margin"]} — поля съедают всю ширину слайда'
                )
            if float(grid["content_bottom"]) <= float(grid["content_y"]):
                problems.append(
                    f'grid_in.content_bottom ({grid["content_bottom"]}) должен быть '
                    f'больше content_y ({grid["content_y"]})'
                )
            header_bottom = float(grid["subtitle_y"]) + float(grid["subtitle_h"])
            if float(grid["title_y"]) + float(grid["title_h"]) > float(grid["subtitle_y"]):
                problems.append("строка заголовка налезает на подзаголовок")
            if header_bottom > float(grid["content_y"]):
                problems.append(
                    f"шапка (до {header_bottom}) налезает на контент "
                    f'(начинается с {grid["content_y"]})'
                )
        except (KeyError, TypeError, ValueError):
            pass

    charts = theme.get("charts")
    if isinstance(charts, dict) and "series" in charts:
        series = charts["series"]
        if not isinstance(series, list) or not series:
            problems.append("charts.series должен быть непустым списком цветов")
        else:
            for i, value in enumerate(series):
                if not HEX.match(str(value)):
                    problems.append(
                        f'charts.series[{i}] = "{value}" — нужен HEX из шести символов'
                    )

    limits = theme.get("limits")
    if isinstance(limits, dict):
        try:
            if float(limits["blocks_max"]) < float(limits["blocks_min"]):
                problems.append("limits.blocks_max меньше limits.blocks_min")
        except (KeyError, TypeError, ValueError):
            pass
        try:
            if float(limits["donut_segments_max"]) < float(limits["donut_segments_min"]):
                problems.append("limits.donut_segments_max меньше limits.donut_segments_min")
        except (KeyError, TypeError, ValueError):
            pass

    return problems


def load_theme(path: Path | None = None) -> dict[str, Any]:
    """Read and validate; raises ThemeError listing every problem found."""
    theme = read_theme(path)
    problems = validate_theme(theme)
    if problems:
        raise ThemeError(
            "pulse/theme.json содержит ошибки:\n  - " + "\n  - ".join(problems)
        )
    return theme


def points(value: Any) -> float:
    """A theme point size in the renderer's own point space."""
    return round(float(value) / SCALE, 1)


def inches(value: Any) -> float:
    return float(value) / SCALE


def main() -> int:
    parser = argparse.ArgumentParser(description="Проверка pulse/theme.json")
    parser.add_argument("--check", action="store_true", help="проверить файл и выйти")
    parser.add_argument("--path", help="путь к theme.json (по умолчанию pulse/theme.json)")
    args = parser.parse_args()

    try:
        theme = read_theme(Path(args.path) if args.path else None)
    except ThemeError as exc:
        print(str(exc), file=sys.stderr)
        return 1

    problems = validate_theme(theme)
    if problems:
        print("pulse/theme.json — найдены ошибки:")
        for problem in problems:
            print(f"  - {problem}")
        print("\nИсправьте их и запустите проверку снова.")
        return 1

    scale = theme["type_scale_pt"]
    grid = theme["grid_in"]
    print("pulse/theme.json — ошибок нет.")
    print(f"  тема      {theme.get('name', 'pulse-library')}")
    print(f"  шрифты    {theme['fonts']['heading']} / {theme['fonts']['body']}")
    print(f"  текст     #{theme['colors']['ink']}   акцент #{theme['colors']['accent']}")
    print(f"  подложка  #{theme['colors']['surface']}   линии #{theme['colors']['line']}")
    print(
        "  кегли     "
        + ", ".join(f"{k} {scale[k]}pt" for k in TYPE_ORDER if k in scale)
    )
    print(
        f"  сетка     поля {grid['margin']}\", заголовок {grid['title_y']}\", "
        f"контент {grid['content_y']}–{grid['content_bottom']}\""
    )
    if isinstance(theme.get("charts"), dict):
        series = theme["charts"].get("series", [])
        print(f"  графики   {len(series)} цветов серий, подписи "
              f"{'вкл' if theme['charts'].get('show_value') else 'выкл'}, "
              f"легенда {'вкл' if theme['charts'].get('show_legend') else 'выкл'}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: templates/pptxgenjs/pulse_theme.js
/*
 * Loads pulse/theme.json and turns it into a renderer preset.
 *
 * The point of this file is that pulse/theme.json is the only place a colour,
 * font, size or margin is written down. Designers edit that; nothing here
 * carries a value of its own, only the conversion.
 *
 * Two coordinate systems are involved. theme.json is written in the units a
 * designer reads off PowerPoint: inches on a 13.333 x 7.5in slide, and points
 * at that size. The renderer authors in a 10 x 5.625in space. Both are 16:9,
 * so the conversion is a single constant and the result is visually identical.
 */

'use strict';

const fs = require('fs');
const path = require('path');

const THEME_PATH = path.join(__dirname, '..', '..', 'pulse', 'theme.json');

// theme.json inches -> renderer inches. 13.333 / 10.
const SCALE = 4 / 3;

const HEX = /^[0-9a-fA-F]{6}$/;

class ThemeError extends Error {}

function readTheme(themePath) {
  const file = themePath || THEME_PATH;
  let raw;
  try {
    raw = fs.readFileSync(file, 'utf8');
  } catch (err) {
    throw new ThemeError(`Не удалось прочитать ${file}: ${err.message}`);
  }
  try {
    return JSON.parse(raw);
  } catch (err) {
    throw new ThemeError(
      `${file} — некорректный JSON: ${err.message}\n` +
      'Частая причина: лишняя запятая перед закрывающей скобкой.',
    );
  }
}

/**
 * Validate the parts the renderer depends on, and say what is wrong in the
 * designer's own terms -- this runs behind `npm run pulse:check`.
 * Returns an array of human-readable problems; empty means the file is good.
 */
function validateTheme(theme) {
  const problems = [];
  const req = (obj, key, where) => {
    if (obj === undefined || obj === null || obj[key] === undefined) {
      problems.push(`${where}: нет обязательного ключа "${key}"`);
      return false;
    }
    return true;
  };

  if (!req(theme, 'colors', 'корень')) return problems;
  for (const key of ['ink', 'ink_muted', 'bg', 'surface', 'accent', 'line']) {
    if (!req(theme.colors, key, 'colors')) continue;
    const value = theme.colors[key];
    if (!HEX.test(String(value))) {
      problems.push(
        `colors.${key} = "${value}" — нужен HEX из шести символов без решётки, ` +
        'например 005AFE',
      );
    }
  }

  if (req(theme, 'fonts', 'корень')) {
    for (const key of ['heading', 'body']) {
      if (req(theme.fonts, key, 'fonts') && !String(theme.fonts[key]).trim()) {
        problems.push(`fonts.${key} пустой`);
      }
    }
  }

  if (req(theme, 'type_scale_pt', 'корень')) {
    for (const key of ['title', 'subtitle', 'body', 'secondary', 'caption', 'micro', 'hero']) {
      if (!req(theme.type_scale_pt, key, 'type_scale_pt')) continue;
      const value = Number(theme.type_scale_pt[key]);
      if (!Number.isFinite(value) || value <= 0) {
        problems.push(`type_scale_pt.${key} = "${theme.type_scale_pt[key]}" — нужно число больше нуля`);
      }
    }
    const scale = theme.type_scale_pt;
    const order = ['micro', 'caption', 'secondary', 'body', 'subtitle', 'title', 'hero'];
    for (let i = 1; i < order.length; i += 1) {
      const lo = Number(scale[order[i - 1]]);
      const hi = Number(scale[order[i]]);
      if (Number.isFinite(lo) && Number.isFinite(hi) && hi < lo) {
        problems.push(
          `type_scale_pt: "${order[i]}" (${hi}) меньше "${order[i - 1]}" (${lo}) — ` +
          'шкала должна расти от micro к hero',
        );
      }
    }
  }

  if (req(theme, 'grid_in', 'корень')) {
    for (const key of ['margin', 'title_y', 'title_h', 'subtitle_y', 'subtitle_h',
      'content_y', 'content_bottom', 'gutter']) {
      if (!req(theme.grid_in, key, 'grid_in')) continue;
      const value = Number(theme.grid_in[key]);
      if (!Number.isFinite(value) || value < 0) {
        problems.push(`grid_in.${key} = "${theme.grid_in[key]}" — нужно число не меньше нуля (дюймы)`);
      }
    }
    const g = theme.grid_in;
    const width = Number((theme.slide && theme.slide.width_in) || 13.333);
    if (Number(g.margin) * 2 >= width) {
      problems.push(`grid_in.margin = ${g.margin} — поля съедают всю ширину слайда`);
    }
    if (Number(g.content_bottom) <= Number(g.content_y)) {
      problems.push(
        `grid_in.content_bottom (${g.content_bottom}) должен быть больше content_y (${g.content_y})`,
      );
    }
    const headerBottom = Number(g.subtitle_y) + Number(g.subtitle_h);
    if (Number(g.title_y) + Number(g.title_h) > Number(g.subtitle_y)) {
      problems.push(
        `строка заголовка (до ${Number(g.title_y) + Number(g.title_h)}) налезает ` +
        `на подзаголовок (начинается с ${g.subtitle_y})`,
      );
    }
    if (headerBottom > Number(g.content_y)) {
      problems.push(
        `шапка (до ${headerBottom}) налезает на контент (начинается с ${g.content_y})`,
      );
    }
  }

  if (theme.charts && theme.charts.series !== undefined) {
    if (!Array.isArray(theme.charts.series) || !theme.charts.series.length) {
      problems.push('charts.series должен быть непустым списком цветов');
    } else {
      theme.charts.series.forEach((value, i) => {
        if (!HEX.test(String(value))) {
          problems.push(`charts.series[${i}] = "${value}" — нужен HEX из шести символов`);
        }
      });
    }
  }

  if (theme.limits) {
    const l = theme.limits;
    if (Number(l.blocks_max) < Number(l.blocks_min)) {
      problems.push('limits.blocks_max меньше limits.blocks_min');
    }
    if (Number(l.donut_segments_max) < Number(l.donut_segments_min)) {
      problems.push('limits.donut_segments_max меньше limits.donut_segments_min');
    }
  }

  return problems;
}

function inches(value) {
  return Number(value) / SCALE;
}

function points(value) {
  return Math.round((Number(value) / SCALE) * 10) / 10;
}

/**
 * Anchors mapping the house type scale onto this theme's steps. The renderer
 * sets font sizes as literals in ~170 places, so a single ramp applied in
 * textOpts() is what moves the whole deck onto the theme's scale while keeping
 * relative hierarchy inside a slide.
 */
function typeRamp(scale) {
  return [
    [7, points(scale.micro)],
    [9, points(scale.caption)],
    [12, points(scale.body)],
    [16, points(scale.subtitle)],
    [26, points(scale.title)],
    [40, points(scale.hero)],
  ];
}

function buildPreset(theme) {
  const c = theme.colors;
  const g = theme.grid_in;
  const t = theme.type_scale_pt;
  const charts = theme.charts || {};

  return {
    bg: c.bg,
    // Kept for presets that invert; this theme never does (light_only below).
    bg_dark: c.ink,
    surface: c.surface,
    surface_strong: c.surface_strong || c.surface,
    text: c.ink,
    text_muted: c.ink_muted,
    text_soft: c.ink_soft || c.ink_muted,
    accent_primary: c.accent,
    accent_secondary: c.accent_bright || c.accent,
    line: c.line,
    font_heading: theme.fonts.heading,
    font_body: theme.fonts.body,

    // Not one slide in the source template has a dark or full-bleed
    // background, so panels tint rather than invert.
    light_only: true,
    square_plates: (theme.plates || {}).square_corners !== false,
    plate_shadow: (theme.plates || {}).shadow === true,

    type_ramp: typeRamp(t),
    readability_contract: {
      min_title_pt: points(t.title),
      min_body_pt: points(t.secondary),
      min_caption_pt: points(t.micro),
      min_support_pt: points(t.secondary),
      min_metadata_pt: points(t.micro),
    },

    content_floor: inches(g.content_bottom),

    chart_series: Array.isArray(charts.series) ? charts.series.slice() : undefined,
    chart_show_value: charts.show_value,
    chart_show_legend: charts.show_legend,
    chart_surface: charts.surface,
    chart_frameless: charts.frameless,

    labels: Object.assign({}, theme.labels),

    // Geometry the pulse renderers read directly, already in renderer inches.
    pulse_grid: {
      margin: inches(g.margin),
      titleY: inches(g.title_y),
      titleH: inches(g.title_h),
      subtitleY: inches(g.subtitle_y),
      subtitleH: inches(g.subtitle_h),
      contentY: inches(g.content_y),
      contentBottom: inches(g.content_bottom),
      gutter: inches(g.gutter),
      blockGap: inches(g.block_gap === undefined ? 0.22 : g.block_gap),
      contentW: inches(Number((theme.slide && theme.slide.width_in) || 13.333) - Number(g.margin) * 2),
      footerY: inches((theme.footer_in && theme.footer_in.y) || 6.74),
      footerH: inches((theme.footer_in && theme.footer_in.height) || 0.42),
      pageNumX: inches((theme.footer_in && theme.footer_in.page_number_x) || 10.88),
      pageNumW: inches((theme.footer_in && theme.footer_in.page_number_w) || 1.77),
    },
    pulse_type: {
      title: points(t.title),
      subtitle: points(t.subtitle),
      body: points(t.body),
      secondary: points(t.secondary),
      caption: points(t.caption),
      micro: points(t.micro),
      hero: points(t.hero),
    },
    pulse_limits: Object.assign({}, theme.limits),
  };
}

let cached = null;

function loadPulsePreset(themePath) {
  if (cached && !themePath) return cached;
  const theme = readTheme(themePath);
  const problems = validateTheme(theme);
  if (problems.length) {
    throw new ThemeError(
      'pulse/theme.json содержит ошибки:\n  - ' + problems.join('\n  - ') +
      '\nЗапустите `npm run pulse:check` для подробностей.',
    );
  }
  const preset = buildPreset(theme);
  if (!themePath) cached = preset;
  return preset;
}

module.exports = {
  THEME_PATH,
  SCALE,
  ThemeError,
  readTheme,
  validateTheme,
  buildPreset,
  loadPulsePreset,
};
